packages/apollo-server-core/src/runHttpQuery.ts

```diff
@@ -97,19 +101,18 @@ export async function runHttpQuery(handlerArguments: Array<any>, request: HttpQu
         }
       }
 
-      // Shallow clone context for queries in batches. This allows
-      // users to distinguish multiple queries in the batch and to
-      // modify the context object without interfering with each other.
-      let context = optionsObject.context;
-      if (isBatch) {
-        context = Object.assign({},  context || {});
+      let context = optionsObject.context || {};
+      if (isFunction(context)) {
+        context = context();
+      } else if (isBatch) {
+        context = Object.assign({}, Object.getPrototypeOf(context), context);
```

Comment:
I think what you'd want here is `Object.assign(Object.create(Object.getPrototypeOf(context)), context)`, because `Object.assign({}, Object.getPrototypeOf(context), context)` doesn't actually set the prototype, it just copies the properties from the prototype.

Reply:
and i thought i knew js, sorry about that